README.md
# RSpec JSON API matchers

# Installation

Add this line to your application's Gemfile:

```ruby
gem "rspec-jsonapi-matchers", github: "Savedo/rspec-jsonapi-matchers", tag: "v0.2.0"
```

## Usage

```ruby
require "rspec/jsonapi/matchers"

describe YourController do
  include RSpec::JsonApi::Matchers

  describe "#index" do
    it "returns error" do
      get :index
      expect(response).to have_jsonapi_error("INVALID")
    end
  end
end
```

## Matchers

have_jsonapi_error:

```ruby
expect(response).to have_jsonapi_error(
  code: "invalid_schema",
  title: "attributes are missing"
)
```

have_jsonapi_resource:
```ruby
expect(response).to have_jsonapi_error("customers", 113", first_name: "John")
```

have_jsonapi_resources:
```ruby
expect(response).to have_jsonapi_resources([
  {
    "id" => "2",
    "type" => "dogs",
    "name" => "Snoopy",
    "owner" => "Charlie"
  },
  {
    "id" => "13",
    "type" => "cats",
    "name" => "Felix"
  }
])
```


CHANGELOG.md
### v0.2.0 (2018-05-09)

* Use `json-schema` instead of `json_schema` dependency (jsmestad)

### v0.1.0 (2016-07-06)

* Public release


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)

require "pry"
require "rspec/jsonapi/matchers"

RSpec.configure do |config|
  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random
end


spec/rspec/jsonapi/matchers_spec.rb
require 'spec_helper'

describe Rspec::Jsonapi::Matchers do
  it 'has a version number' do
    expect(Rspec::Jsonapi::Matchers::VERSION).not_to be nil
  end

  it 'does something useful' do
    expect(false).to eq(true)
  end
end


spec/rspec/jsonapi/matchers/have_jsonapi_error_matcher_spec.rb
require "spec_helper"

describe RSpec::JsonApi::Matchers::HaveJsonApiErrorMatcher do
  include RSpec::JsonApi::Matchers

  def fail_with(message)
    raise_error(RSpec::Expectations::ExpectationNotMetError, message)
  end

  let(:response) { double("Response", body: body) }

  context "when response is not JSON" do
    let(:body) { "none-json" }

    it "fails" do
      expect do
        expect(response).to have_jsonapi_error(code: "test", title: "Test")
      end.to fail_with(/Response contains malformed JSON/)
    end
  end

  context 'when response does not contain "errors" item' do
    let(:body) do
      {
        "data" => {
          "id" => "123",
          "type" => "foo",
          "attributes" => {
            "foo" => "bar"
          }
        }
      }.to_json
    end

    it "fails" do
      expect do
        expect(response).to have_jsonapi_error(code: "test", title: "Test")
      end.to fail_with(/Response does not contain "errors" item/)
    end
  end

  context "when response does not contain matching error" do
    let(:body) do
      { "errors" => [] }.to_json
    end

    it "fails" do
      expect do
        expect(response).to have_jsonapi_error(code: "test", title: "Test")
      end.to fail_with(/expected to have error/i)
    end
  end

  context "when response contain matching error" do
    let(:body) do
      {
        "errors" => [{ "code" => "test", "title" => "Test" }]
      }.to_json
    end

    it "passes" do
      expect(response).to have_jsonapi_error(code: "test", title: "Test")
    end
  end
end


lib/rspec/jsonapi/matchers/base_matcher.rb
module RSpec
  module JsonApi
    module Matchers
      # Basic class for JSON API matchers. Subclasses must implement +verify!+ method
      # that raises Api::ExpectationFailure with meaningful message in case of failure.
      class BaseMatcher
        attr_reader :response, :parsed_response

        # @param response [ActionDispatch::Response]
        #
        # @return [Boolean]
        def matches?(response)
          @response = response
          set_parsed_response!
          validate_schema!
          verify!
          true
        rescue ExpectationFailure => err
          @message = "#{err.message}.\n#{pretty_response}"
          false
        end

        def failure_message
          @message || "expected to #{description}"
        end

        def failure_message_when_negated
          @message || "expected not to #{description}"
        end

        # Default description.
        # E.g. for HaveJsonapiResource => "have jsonapi resource"
        def description
          self.class.name.underscore.humanize.downcase
        end

        private def set_parsed_response!
          @parsed_response = JSON.parse(response.body)
        rescue JSON::ParserError
          raise(ExpectationFailure , "Response contains malformed JSON")
        end

        private def schema
          @schema ||= begin
            parsed_schema = JSON.parse(File.read(File.join(__dir__,"/jsonapi.json")))
            JsonSchema.parse!(parsed_schema).tap do |s|
              s.expand_references!
            end
          end
        rescue JSON::ParserError => error
          raise ExpectationFailure, "JSON Schema is not valid JSON. #{error&.message}"
        rescue JsonSchema::SchemaError => error
          raise ExpectationFailure, "JSON Schema is valid JSON, but is not a valid JSON Schema. #{error&.message}"
        end

        private def validate_schema!
          schema.validate!(parsed_response)
        rescue JsonSchema::Error => error
          msg = "Response does not match JSON API schema. #{error.message}"
          raise(ExpectationFailure, msg)
        end

        private def pretty_response
          if parsed_response
            JSON.pretty_generate(parsed_response)
          else
            response.body
          end
        end
      end
    end
  end
end


lib/rspec/jsonapi/matchers/have_jsonapi_error_matcher.rb
module RSpec
  module JsonApi
    module Matchers
      class HaveJsonApiErrorMatcher < BaseMatcher
        attr_reader :expected_properties, :errors

        def initialize(expected_properties)
          @expected_properties = expected_properties.stringify_keys
        end

        private def verify!
          set_errors!
          verify_expected_properties!
        end

        private def set_errors!
          if parsed_response["errors"]
            @errors = parsed_response["errors"]
          else
            raise(ExpectationFailure, "Response does not contain \"errors\" item")
          end
        end

        private def verify_expected_properties!
          detected_error = errors.detect do |err|
            expected_properties.all? { |key, val| err[key] == val }
          end
          raise(ExpectationFailure, "Expected to have error that matches #{expected_properties.inspect}") unless detected_error
        end
      end
    end
  end
end


lib/rspec/jsonapi/matchers/have_jsonapi_resource_matcher.rb
module RSpec
  module JsonApi
    module Matchers
      class HaveJsonApiResourceMatcher < BaseMatcher
        attr_reader :data, :expected_resource

        def initialize(attributes)
          attrs = attributes.stringify_keys
          type = attrs.delete("type")
          id = attrs.delete("id")
          @expected_resource = Resource.new(type, id, attrs)
        end

        private def verify!
          set_data!
          actual_resource = Resource.new(data["type"], data["id"] , data["attributes"])
          verify_resource!(expected_resource, actual_resource)
        end

        private def set_data!
          if parsed_response["data"]
            @data = parsed_response["data"]
          else
            raise(ExpectationFailure , "Response does not contain \"data\" property")
          end
        end

        private def verify_resource!(expected_resource, actual_resource)
          verify_type!(expected_resource, actual_resource)
          verify_id!(expected_resource, actual_resource)
          verify_attrs!(expected_resource, actual_resource)
        end

        private def verify_type!(expected_resource, actual_resource)
          return unless expected_resource.type.present?
          return if actual_resource.type.to_s == expected_resource.type.to_s
          raise(ExpectationFailure , "Expected resource type to be #{expected_resource.type.inspect}, but got #{actual_resource.type.inspect}")
        end

        private def verify_id!(expected_resource, actual_resource)
          return unless expected_resource.id.present?
          return if actual_resource.id.to_s == expected_resource.id.to_s
          raise(ExpectationFailure , "Expected resource id to be #{expected_resource.id.inspect}, but got #{actual_resource.id.inspect}")
        end

        private def verify_attrs!(expected_resource, actual_resource)
          return unless expected_resource.attributes.present?
          raise(ExpectationFailure, "attributes are not present in the response") unless actual_resource.attributes

          expected_resource.attributes.each do |attr, expected_value|
            actual_value = actual_resource.attributes[attr]
            if actual_value != expected_value
              raise(ExpectationFailure, "Expected attribute #{attr.inspect} to be #{expected_value.inspect}, but got #{actual_value.inspect}")
            end
          end
        end
      end
    end
  end
end


lib/rspec/jsonapi/matchers/have_jsonapi_resources_matcher.rb
module RSpec
  module JsonApi
    module Matchers
      class HaveJsonApiResourcesMatcher < HaveJsonApiResourceMatcher
        attr_reader :array_of_expected_attrs

        def initialize(array_of_expected_attrs)
          if array_of_expected_attrs.is_a?(Array)
            @array_of_expected_attrs = array_of_expected_attrs
          else
            raise(ArgumentError, "have_jsonapi_resources matcher accepts Array of Hashes, got #{array_of_expected_attrs.inspect}")
          end
        end

        def verify!
          set_data!
          array_of_expected_attrs.each do |original_attrs|
            attrs = original_attrs.stringify_keys
            type = attrs.delete("type")
            id = attrs.delete("id")
            verify_expected_resource!(type, id, attrs)
          end
        end

        private def verify_expected_resource!(expected_type, expected_id, expected_attrs)
          expected_resource = Resource.new(expected_type, expected_id, expected_attrs)
          match = data.any? do |actual_resource_data|
            match_resource?(expected_resource, actual_resource_data)
          end
          return if match

          raise(ExpectationFailure,
            "Response does not contain expected resource with type=#{expected_type.inspect}, "\
            "id=#{expected_id}, attributes=#{expected_attrs.inspect}"
          )
        end

        private def match_resource?(expected_resource, actual_data)
          actual_resource = Resource.new(actual_data["type"], actual_data["id"], actual_data["attributes"])
          verify_resource!(expected_resource, actual_resource)
          true
        rescue ExpectationFailure
          false
        end
      end
    end
  end
end


lib/rspec/jsonapi/matchers.rb
require "active_support"
require "active_support/core_ext/hash"
require "json_schema"

require "rspec/jsonapi/matchers/version"
require "rspec/jsonapi/matchers/resource"
require "rspec/jsonapi/matchers/base_matcher"
require "rspec/jsonapi/matchers/have_jsonapi_error_matcher"
require "rspec/jsonapi/matchers/have_jsonapi_resource_matcher"
require "rspec/jsonapi/matchers/have_jsonapi_resources_matcher"

module RSpec
  module JsonApi
    module Matchers
      class ExpectationFailure < ::StandardError; end

      # Verify that response contains error in JSON API format with given
      # code and title.
      # For more info please check: http://jsonapi.org/format/1.0/#error-objects
      #
      # @example
      #   expect(response).to have_jsonapi_error(
      #     code: "invalid_schema",
      #     title: "attributes are missing"
      #   )
      #
      # @param expected_properties [Hash]
      def have_jsonapi_error(expected_properties)
        HaveJsonApiErrorMatcher.new(expected_properties)
      end

      # Verify that response contains resource with given type, id and attributes
      # in JSON API format.
      # For more info please check: http://jsonapi.org/format/#document-resource-objects
      #
      # @example
      #   expect(response).to have_jsonapi_error("customers", 113", first_name: "John")
      #
      # @param attrs [Hash]
      def have_jsonapi_resource(attrs = {})
        HaveJsonApiResourceMatcher.new(attrs)
      end

      # Verify that response contains JSON API resources with parameters.
      #
      # @example
      #   expect(response).to have_jsonapi_resources([
      #     {
      #       "id" => "2",
      #       "type" => "dogs",
      #       "name" => "Snoopy",
      #       "owner" => "Charlie"
      #     },
      #     {
      #       "id" => "13",
      #       "type" => "cats",
      #       "name" => "Felix"
      #     }
      #   ])
      # @param array_of_expected_attrs [Array<Hash>]
      def have_jsonapi_resources(array_of_expected_attrs)
        HaveJsonApiResourcesMatcher.new(array_of_expected_attrs)
      end
    end
  end
end
